### PR TITLE
Implement preferred_chunks for netcdf 4 backends

### LINKED ISSUE #1440: If a NetCDF file is chunked on disk, open it with compatible dask chunks
NetCDF4 data can be saved as chunks on disk, which [has several benefits](https://www.unidata.ucar.edu/blogs/developer/entry/chunking_data_why_it_matters) including efficient reads when using a compatible chunk shape.  This is particularly important for files with chunk-based compression (ie all nc4 files with compression) or on HPC and parallel file systems ([eg](http://anusf.anu.edu.au/~jxa900/pres/COMP4300-2015/jxa900-Parallel-IO-reduced.pdf)), where IO is typically dominated by the number of reads and chunks-from-disk are often cached.  Caches are also common in network data backends such as Thredds OPeNDAP, in which case using disk-compatible chunks will reduce cache pressure as well as latency.

Xarray *can* use chunks, of course, but as of v0.9 the chunk size has to be specified manually - and the easiest way to discover it is to open the file and look at the `_Chunksizes` attribute for each variable.  I propose that `xr.open_dataset` (and `array`, and `mfdataset`) change their default behaviour.

If Dask is available and `chunks=None` (the default), `chunks` should be taken from the file on disk.  This may lead to a chunked or unchunked dataset.  To force an un-chunked load, users can specify `chunks={}`, or simple `.load()` the dataset after opening it.

### PR DESCRIPTION
**Pull request by mraspaud**
According to the `open_dataset` documentation, using `chunks="auto"` or `chunks={}` should yield datasets with variables chunked depending on the preferred chunks of the backend. However neither the netcdf4 nor the h5netcdf backend seem to implement the `preferred_chunks` encoding attribute needed for this to work.

This PR adds this attribute to the encoding upon data reading. This results in `chunks="auto"` in `open_dataset` returning variables with chunk sizes multiples of the chunks in the nc file, and for `chunks={}`, returning the variables with then exact nc chunk sizes.  

- [x] Closes #1440
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

**Comment by mraspaud** [comment]
Hope this is up to the standards! Please tell me if there is anything I can do to improve this PR.

**Comment by mraspaud** [comment]
@djhoese if `chunks={}`, the same chunks size as the on-disk file will be used. if `chunks="auto"`, a multiple of the on-disk chunk size will be used to accommodate the dask chunk size preferences. This is in line with what is said in the `open_dataset` docstring as I understand it.
For our case, using `chunks="auto"` with this patch resulted in a significant performance boost in processing the data as compared to the main branch.

File: xarray/tests/test_backends.py [hunk 1]
```diff
@@ -1535,6 +1535,14 @@ def test_keep_chunksizes_if_no_original_shape(self) -> None:
...
             )
 
+    def test_preferred_chunks_is_present(self) -> None:
+        ds = Dataset({"x": [1, 2, 3]})
+        chunksizes = (2,)
+        ds.variables["x"].encoding = {"chunksizes": chunksizes}
+
+        with self.roundtrip(ds) as actual:
+            assert actual["x"].encoding["preferred_chunks"] == {"x": 2}
+
```
Comment (jhamman) [inline xarray/tests/test_backends.py:1545]: This test is a good start. However, I would like to see a more thorough set of tests that set up NetCDF files with various chunk schemas and confirm that the `auto`/`{}` options return expected outputs after loading.

**Review by jhamman — COMMENTED** [review]
Thanks for taking this on @mraspaud. This has been something folks have wanted for some time now!

File: xarray/tests/test_backends.py (hunk 1 above)
Comment (mraspaud) [reply xarray/tests/test_backends.py:1545]: Ok, I'll do that. Is this file the right place for these tests?

Comment (mraspaud) [reply xarray/tests/test_backends.py:1545]: @jhamman done, tell me what you think :)

File: xarray/tests/test_backends.py [hunk 2]
```diff
@@ -1535,6 +1535,66 @@ def test_keep_chunksizes_if_no_original_shape(self) -> None:
...
+        with create_tmp_file() as tmp_file:
+            with nc4.Dataset(tmp_file, mode="w") as nc:
+                nc.createDimension("t", t_size)
+                nc.createDimension("x", x_size)
+                nc.createDimension("y", y_size)
+                nc.createVariable("image", "int16", ("t", "y", "x"), fill_value=-1,
+                                  chunksizes=(t_chunksize, y_chunksize, x_chunksize))
+                v = nc.variables["image"]
+                v[:] = np.arange(t_size * x_size * y_size).reshape((t_size, y_size, x_size))
```
Comment (jhamman) [inline xarray/tests/test_backends.py:1584]: While is this a fine way to create these test files, I would rather see us use xarray's backends here. The [`chunksizes`](https://docs.xarray.dev/en/stable/user-guide/io.html#chunk-based-compression) encoding option can be used to specify encoded chunksize in the netcdf4/h5netcdf engines.

Comment (mraspaud) [reply xarray/tests/test_backends.py]: Good idea, done in 29c93b8

**Comment by dcherian** [comment]
```
Found 5 errors in 1 file (checked 142 source files)
xarray/tests/test_backends.py:1568: error: Need type annotation for "tmp_file"  [var-annotated]
xarray/tests/test_backends.py:1577: error: Argument 1 to "contextmanager" has incompatible type "Callable[[NetCDF4Base, Any, Any], None]"; expected "Callable[[NetCDF4Base, Any, Any], Iterator[<nothing>]]"  [arg-type]
xarray/tests/test_backends.py:1578: error: The return type of a generator function should be "Generator" or one of its supertypes  [misc]
xarray/tests/test_backends.py:1599: error: Need type annotation for "tmp_file"  [var-annotated]
```

@Illviljan @headtr1ck  help!

File: xarray/tests/test_backends.py [hunk 3]
```diff
@@ -1535,6 +1535,77 @@ def test_keep_chunksizes_if_no_original_shape(self) -> None:
...
+            (1, y_size, x_size), (1, y_chunksize, x_chunksize)
+        ) as tmp_file:
+            with open_dataset(tmp_file, chunks={}, engine=self.engine) as ds:
+                for chunksizes, expected in zip(
+                    ds["image"].data.chunks, (1, y_chunksize, x_chunksize)
+                ):
+                    assert all(np.asanyarray(chunksizes) == expected)
+
+    @contextlib.contextmanager
+    def create_chunked_file(self, array_shape, chunk_sizes) -> None:
```
Comment (headtr1ck) [inline xarray/tests/test_backends.py:1578]:
```suggestion
    def create_chunked_file(self, array_shape: tuple[int, int, int], chunk_sizes: tuple[int, int, int]) -> Generator[str, None, None]:
```
I haven't checked it though, but it should bring you in the right direction.

**Comment by mraspaud** [comment]
I attempted to fix the type annotation problem, please tell me if there is more.

**Review by dcherian — APPROVED** [review]
LGTM! Thanks for your patience here @mraspaud 

Can you add a note to `whats-new` to advertise this  great improvement please?

File: xarray/tests/test_backends.py [hunk 4]
```diff
@@ -1535,6 +1535,79 @@ def test_keep_chunksizes_if_no_original_shape(self) -> None:
...
+            (1, y_size, x_size), (1, y_chunksize, x_chunksize)
+        ) as tmp_file:
+            with open_dataset(tmp_file, chunks={}, engine=self.engine) as ds:
+                for chunksizes, expected in zip(
+                    ds["image"].data.chunks, (1, y_chunksize, x_chunksize)
+                ):
+                    assert all(np.asanyarray(chunksizes) == expected)
+
+    @contextlib.contextmanager
+    def create_chunked_file(
```
Comment (dcherian) [inline xarray/tests/test_backends.py:1578]: Is it possible to pass `chunksizes` to `roundtrip` instead?

Comment (mraspaud) [reply xarray/tests/test_backends.py]: good idea, done

**Comment by mraspaud** [comment]
What's new now updated

**Comment by dcherian** [comment]
The mypy failures are. related to these changes I think:
```
xarray/tests/test_backends.py:1559: error: "str" has no attribute "data"  [attr-defined]
xarray/tests/test_backends.py:1559: error: Invalid index type "str" for "str"; expected type "SupportsIndex | slice"  [index]
xarray/tests/test_backends.py:1576: error: "str" has no attribute "data"  [attr-defined]
xarray/tests/test_backends.py:1576: error: Invalid index type "str" for "str"; expected type "SupportsIndex | slice"  [index]
xarray/tests/test_backends.py:1610: error: "str" has no attribute "encoding"  [attr-defined]
xarray/tests/test_backends.py:1610: error: Invalid index type "str" for "str"; expected type "SupportsIndex | slice"  [index]
```

@Illviljan can you take a look please?

File: xarray/tests/test_backends.py [hunk 5]
```diff
@@ -1536,6 +1536,83 @@ def test_keep_chunksizes_if_no_original_shape(self) -> None:
...
+            ):
+                assert all(np.asanyarray(chunksizes) == expected)
+
+    @contextlib.contextmanager
+    def chunked_roundtrip(
+        self,
+        array_shape: tuple[int, int, int],
+        chunk_sizes: tuple[int, int, int],
+        open_kwargs: dict[str, Any] | None = None,
+    ) -> Generator[str, None, None]:
```
Comment (Illviljan) [inline xarray/tests/test_backends.py:1586]:
```suggestion
    ) -> Generator[Dataset]:
```

File: xarray/tests/test_backends.py [hunk 6]
```diff
@@ -1536,6 +1536,83 @@ def test_keep_chunksizes_if_no_original_shape(self) -> None:
...
+            ):
+                assert all(np.asanyarray(chunksizes) == expected)
+
+    @contextlib.contextmanager
+    def chunked_roundtrip(
+        self,
+        array_shape: tuple[int, int, int],
+        chunk_sizes: tuple[int, int, int],
+        open_kwargs: dict[str, Any] | None = None,
+    ) -> Generator[Dataset]:
```
Comment (Illviljan) [inline xarray/tests/test_backends.py:1586]:
```suggestion
    ) -> Generator[Dataset, None, None]:
```

**Comment by Illviljan** [comment]
Thanks, @mraspaud !

**Comment by mraspaud** [comment]
My pleasure! thanks for merging!

**Pull request merged**